// ngx-typeahead/esm5/ngx-typeahead.js
/**
 * @license ngx-typeahead
 * MIT license
 */
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, HostListener, Input, NgModule, Output, ViewChild, ViewContainerRef } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClient, HttpClientJsonpModule, HttpClientModule, HttpParams } from '@angular/common/http';
import 'rxjs/add/operator/debounceTime';
import 'rxjs/add/operator/concat';
import 'rxjs/add/operator/distinctUntilChanged';
import 'rxjs/add/operator/share';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/switchMap';
import 'rxjs/add/observable/fromEvent';
import { Observable as Observable$1 } from 'rxjs/Observable';
import { __assign } from 'tslib';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {number} */
var Key = {
    Backspace: 8,
    Tab: 9,
    Enter: 13,
    Shift: 16,
    Escape: 27,
    ArrowLeft: 37,
    ArrowRight: 39,
    ArrowUp: 38,
    ArrowDown: 40,
    // http://unixpapa.com/js/key.html
    MacCommandLeft: 91,
    MacCommandRight: 93,
    MacCommandFirefox: 224,
};
Key[Key.Backspace] = "Backspace";
Key[Key.Tab] = "Tab";
Key[Key.Enter] = "Enter";
Key[Key.Shift] = "Shift";
Key[Key.Escape] = "Escape";
Key[Key.ArrowLeft] = "ArrowLeft";
Key[Key.ArrowRight] = "ArrowRight";
Key[Key.ArrowUp] = "ArrowUp";
Key[Key.ArrowDown] = "ArrowDown";
Key[Key.MacCommandLeft] = "MacCommandLeft";
Key[Key.MacCommandRight] = "MacCommandRight";
Key[Key.MacCommandFirefox] = "MacCommandFirefox";

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @param {?} keyCode
 * @return {?}
 */
function validateNonCharKeyCode(keyCode) {
    return [
        Key.Enter,
        Key.Tab,
        Key.Shift,
        Key.ArrowLeft,
        Key.ArrowUp,
        Key.ArrowRight,
        Key.ArrowDown,
        Key.MacCommandLeft,
        Key.MacCommandRight,
        Key.MacCommandFirefox
    ].every(function (codeKey) { return codeKey !== keyCode; });
}
/**
 * @param {?} keyCode
 * @return {?}
 */
function validateArrowKeys(keyCode) {
    return keyCode === Key.ArrowDown || keyCode === Key.ArrowUp;
}
/**
 * @param {?} index
 * @param {?} currentIndex
 * @return {?}
 */
function isIndexActive(index, currentIndex) {
    return index === currentIndex;
}
/**
 * @param {?} query
 * @param {?=} queryParamKey
 * @param {?=} customParams
 * @return {?}
 */
function createParamsForQuery(query, queryParamKey, customParams) {
    if (queryParamKey === void 0) { queryParamKey = 'q'; }
    if (customParams === void 0) { customParams = {}; }
    var /** @type {?} */ searchParams = __assign((_a = {}, _a[queryParamKey] = query, _a), customParams);
    // tslint:disable-next-line
    var /** @type {?} */ setParam = function (acc, param) { return acc.set(param, searchParams[param]); };
    var /** @type {?} */ params = Object.keys(searchParams).reduce(setParam, new HttpParams());
    return params;
    var _a;
}
/**
 * @param {?=} method
 * @return {?}
 */
function resolveApiMethod(method) {
    if (method === void 0) { method = ''; }
    var /** @type {?} */ isMethodValid = [
        'get',
        'post',
        'put',
        'delete',
        'patch',
        'request'
    ].some(function (methodName) { return method === methodName; });
    var /** @type {?} */ apiMethod = isMethodValid ? method : 'get';
    return apiMethod;
}
/**
 * @param {?} currentIndex
 * @param {?} stepUp
 * @return {?}
 */
function resolveNextIndex(currentIndex, stepUp) {
    var /** @type {?} */ step = stepUp ? 1 : -1;
    var /** @type {?} */ topLimit = 9;
    var /** @type {?} */ bottomLimit = 0;
    var /** @type {?} */ currentResultIndex = currentIndex + step;
    var /** @type {?} */ resultIndex = currentResultIndex;
    if (currentResultIndex === topLimit + 1) {
        resultIndex = bottomLimit;
    }
    if (currentResultIndex === bottomLimit - 1) {
        resultIndex = topLimit;
    }
    return resultIndex;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var NgxTypeAheadComponent = (function () {
    function NgxTypeAheadComponent(element, viewContainer, http$$1, cdr) {
        this.element = element;
        this.viewContainer = viewContainer;
        this.http = http$$1;
        this.cdr = cdr;
        this.showSuggestions = false;
        this.taUrl = '';
        this.taParams = {};
        this.taQueryParam = 'q';
        this.taApi = 'jsonp';
        this.taApiMethod = 'get';
        this.taList = [];
        this.taSelected = new EventEmitter();
        this.suggestionIndex = 0;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.handleEsc = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.keyCode === Key.Escape) {
            this.hideSuggestions();
            event.preventDefault();
        }
    };
    /**
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ onkeyDown$ = this.onElementKeyDown();
        this.subscriptions = [
            this.filterEnterEvent(onkeyDown$),
            this.listenAndSuggest(),
            this.navigateWithArrows(onkeyDown$)
        ];
        this.renderTemplate();
    };
    /**
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subscriptions.forEach(function (sub) { return sub.unsubscribe(); });
        this.subscriptions.length = 0;
    };
    /**
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.renderTemplate = /**
     * @return {?}
     */
    function () {
        this.viewContainer.createEmbeddedView(this.suggestionsTplRef);
        this.cdr.markForCheck();
    };
    /**
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.onElementKeyDown = /**
     * @return {?}
     */
    function () {
        return Observable$1.fromEvent(this.element.nativeElement, 'keydown').share();
    };
    /**
     * @param {?} elementObs
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.filterEnterEvent = /**
     * @param {?} elementObs
     * @return {?}
     */
    function (elementObs) {
        var _this = this;
        return elementObs
            .filter(function (e) { return e.keyCode === Key.Enter; })
            .subscribe(function (event) {
            event.preventDefault();
            _this.handleSelectSuggestion(_this.activeResult);
        });
    };
    /**
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.listenAndSuggest = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Observable$1.fromEvent(this.element.nativeElement, 'keyup')
            .filter(function (e) { return validateNonCharKeyCode(e.keyCode); })
            .map(function (e) { return e.target.value; })
            .debounceTime(300)
            .concat()
            .distinctUntilChanged()
            .filter(function (query) { return query.length > 0; })
            .switchMap(function (query) { return _this.suggest(query); })
            .subscribe(function (results) {
            _this.results = results;
            _this.showSuggestions = true;
            _this.suggestionIndex = 0;
            _this.cdr.markForCheck();
        });
    };
    /**
     * @param {?} elementObs
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.navigateWithArrows = /**
     * @param {?} elementObs
     * @return {?}
     */
    function (elementObs) {
        var _this = this;
        return elementObs
            .filter(function (e) { return validateArrowKeys(e.keyCode); })
            .map(function (e) { return e.keyCode; })
            .subscribe(function (keyCode) {
            _this.suggestionIndex = resolveNextIndex(_this.suggestionIndex, keyCode === Key.ArrowDown);
            _this.showSuggestions = true;
            _this.cdr.markForCheck();
        });
    };
    /**
     * @param {?} query
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.suggest = /**
     * @param {?} query
     * @return {?}
     */
    function (query) {
        return this.taList.length
            ? this.createListSource(this.taList)
            : this.request(query);
    };
    /**
     * peforms a jsonp/http request to search with query and params
     * @param query the query to search from the remote source
     */
    /**
     * peforms a jsonp/http request to search with query and params
     * @param {?} query the query to search from the remote source
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.request = /**
     * peforms a jsonp/http request to search with query and params
     * @param {?} query the query to search from the remote source
     * @return {?}
     */
    function (query) {
        var /** @type {?} */ url = this.taUrl;
        var /** @type {?} */ searchConfig = createParamsForQuery(query, this.taQueryParam, this.taParams);
        var /** @type {?} */ options = {
            params: searchConfig
        };
        var /** @type {?} */ isJsonpApi = this.taApi === 'jsonp';
        return isJsonpApi
            ? this.requestJsonp(url, options, this.taCallbackParamValue)
            : this.requestHttp(url, options);
    };
    /**
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.requestHttp = /**
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    function (url, options) {
        var /** @type {?} */ apiMethod = resolveApiMethod(this.taApiMethod);
        var /** @type {?} */ responseTransformMethod = this.taResponseTransform || (function (item) { return item; });
        return this.http[apiMethod](url, options)
            .map(function (results) { return results.map(responseTransformMethod); });
    };
    /**
     * @param {?} url
     * @param {?} options
     * @param {?=} callback
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.requestJsonp = /**
     * @param {?} url
     * @param {?} options
     * @param {?=} callback
     * @return {?}
     */
    function (url, options, callback) {
        if (callback === void 0) { callback = 'callback'; }
        var /** @type {?} */ params = options.params.toString();
        return this.http.jsonp(url + "?" + params, callback)
            .map(function (response) { return response[1]; })
            .map(function (results) { return results.map(function (result) { return result[0]; }); });
    };
    /**
     * @param {?} index
     * @param {?} result
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.markIsActive = /**
     * @param {?} index
     * @param {?} result
     * @return {?}
     */
    function (index, result) {
        var /** @type {?} */ isActive = isIndexActive(index, this.suggestionIndex);
        if (isActive) {
            this.activeResult = result;
        }
        return isActive;
    };
    /**
     * @param {?} suggestion
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.handleSelectSuggestion = /**
     * @param {?} suggestion
     * @return {?}
     */
    function (suggestion) {
        this.hideSuggestions();
        this.taSelected.emit(suggestion);
    };
    /**
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.hideSuggestions = /**
     * @return {?}
     */
    function () {
        this.showSuggestions = false;
    };
    /**
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.hasItemTemplate = /**
     * @return {?}
     */
    function () {
        return this.taItemTpl !== undefined;
    };
    /**
     * @param {?} list
     * @return {?}
     */
    NgxTypeAheadComponent.prototype.createListSource = /**
     * @param {?} list
     * @return {?}
     */
    function (list) {
        return new Observable$1(function (observer) {
            observer.next(list);
        });
    };
    NgxTypeAheadComponent.decorators = [
        { type: Component, args: [{
                    selector: '[ngxTypeahead]',
                    styles: [
                        "\n  .results {\n    position: absolute;\n  }\n  .typeahead-backdrop {\n    bottom: 0;\n    left: 0;\n    position: fixed;\n    right: 0;\n    top: 0;\n    z-index: 1;\n  }\n  .list-group-item {\n    position: relative;\n    z-index: 2;\n  }\n  "
                    ],
                    template: "\n  <ng-template #suggestionsTplRef>\n  <section class=\"list-group results\" *ngIf=\"showSuggestions\">\n    <div class=\"typeahead-backdrop\" (click)=\"hideSuggestions()\"></div>\n    <button type=\"button\" class=\"list-group-item\"\n      *ngFor=\"let result of results; let i = index;\"\n      [class.active]=\"markIsActive(i, result)\"\n      (click)=\"handleSelectSuggestion(result)\">\n      <span *ngIf=\"!taItemTpl\"><i class=\"fa fa-search\"></i> {{ result }}</span>\n      <ng-template\n        [ngTemplateOutlet]=\"taItemTpl\"\n        [ngTemplateOutletContext]=\"{ $implicit: {result: result, index: i} }\"\n      ></ng-template>\n    </button>\n  </section>\n  </ng-template>\n  "
                },] },
    ];
    /** @nocollapse */
    NgxTypeAheadComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: ViewContainerRef, },
        { type: HttpClient, },
        { type: ChangeDetectorRef, },
    ]; };
    NgxTypeAheadComponent.propDecorators = {
        "taItemTpl": [{ type: Input },],
        "taUrl": [{ type: Input },],
        "taParams": [{ type: Input },],
        "taQueryParam": [{ type: Input },],
        "taCallbackParamValue": [{ type: Input },],
        "taApi": [{ type: Input },],
        "taApiMethod": [{ type: Input },],
        "taResponseTransform": [{ type: Input },],
        "taList": [{ type: Input },],
        "taSelected": [{ type: Output },],
        "suggestionsTplRef": [{ type: ViewChild, args: ['suggestionsTplRef',] },],
        "handleEsc": [{ type: HostListener, args: ['keydown', ['$event'],] },],
    };
    return NgxTypeAheadComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var NgxTypeaheadModule = (function () {
    function NgxTypeaheadModule() {
    }
    NgxTypeaheadModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [NgxTypeAheadComponent],
                    exports: [NgxTypeAheadComponent, CommonModule],
                    imports: [CommonModule, HttpClientModule, HttpClientJsonpModule],
                    providers: []
                },] },
    ];
    /** @nocollapse */
    NgxTypeaheadModule.ctorParameters = function () { return []; };
    return NgxTypeaheadModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Angular library starter.
 * Build an Angular library compatible with AoT compilation & Tree shaking
 * Copyright Roberto Simonetti.
 * MIT license.
 * https://github.com/robisim74/angular-library-starter
 */
/**
 * Entry point for all public APIs of the package.
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { NgxTypeaheadModule, NgxTypeAheadComponent as ɵa };
